codeworld-compiler/src/ParseCode.hs

```diff
@@ -0,0 +1,128 @@
+checkParsedCode src err = do
+    result <- parseFile src
+    let parseList = words $ show result
+        parseresult = head parseList
+    if parseresult == "ParseFailed" then do
+        source <- readFile src
+        let sourceSplitList = splitOn "\n" source
+            errLin = errLineLocation parseList
+            errCol = errColumnLocation parseList
+        B.writeFile err (pack ("ParseError, Error at: program.hs:"
+                ++ (show $ errLin + 1) ++ ":" ++ (show errCol)
+                ++ "\n>  " ++ (sourceSplitList !! errLin)))
+        return False
+        else do
+            let outp = getErrors (getSrcSpanInfo result)
+            if outp == [] then return True
+                else do
+                    let outpString = tail $ tail $ init $ init (show outp)
+                    B.writeFile err (pack outpString)
+                    return False
+
+getSrcSpanInfo :: ParseResult (Module SrcSpanInfo) -> Module SrcSpanInfo
+getSrcSpanInfo (ParseOk x) = x
+
+getErrors :: Module SrcSpanInfo -> [String]
+getErrors m = everything (++) (mkQ [] badExps) m ++
+              everything (++) (mkQ [] badMatches) m
+              -- everything (++) (mkQ [] badInfixExps) m
+
+{-
+badInfixExps :: Exp SrcSpanInfo -> [String]
+badInfixExps (InfixApp a (App l _ _) c d) = ["Error at InfixApp: Missing &"
+        ++ "program.hs:"
+        ++ (show $ parseSrcSpanStartLine a)
+        ++ ":"
+        ++ (show $ parseSrcSpanStartColumn a)]
+badInfixExps (InfixApp a b c (App l _ _)) = ["Error at InfixApp: Missing &"
+        ++ "program.hs:"
+        ++ (show $ parseSrcSpanStartLine l)
+        ++ ":"
+        ++ (show $ parseSrcSpanStartColumn l)]
+badInfixExps (InfixApp a b c d) | not (isGoodInfixExpRhs c) = ["Error at InfixApp: program.hs:"
+        ++ (show $ parseSrcSpanStartLine a)
+        ++ ":"
+        ++ (show $ parseSrcSpanStartColumn a)]
+badInfixExps _                                              = []
+-}
+
+{-
+badExps (App x (Lit m _) (Lit m _))  = []
+badExps (App x (Lit m _) (Var m _))  = []
+-}
+badExps :: Exp SrcSpanInfo -> [String]
+badExps (App x (Var l _) e) | not (isGoodExpRhs e) = [
+        "Missing parantheses in function applicaion at: program.hs:"
+        ++ (show $ parseSrcSpanEndLine l) ++ ":"
+        ++ (show $ parseSrcSpanEndColumn l)]
+badExps _                                  = []
+
+badMatches :: Match SrcSpanInfo -> [String]
+badMatches (Match _ _ pats _ _) = [ "Bad pattern match!" | p <- pats, not (isGoodPatRhs p) ]
+badMatches _                    = []
+
+isGoodExpRhs :: Exp l -> Bool
+isGoodExpRhs (Paren _ _)   = True
+isGoodExpRhs (Tuple _ _ _) = True
+isGoodExpRhs _             = False
+
+isGoodInfixExpRhs :: QOp l -> Bool
+isGoodInfixExpRhs (QVarOp _ _) = True
+isGoodInfixExpRhs _            = False
+
+isGoodPatRhs :: Pat l -> Bool
+isGoodPatRhs (PParen _ _)   = True
+isGoodPatRhs (PTuple _ _ _) = True
+isGoodPatRhs _              = False
+
+parseSrcSpanEndLine :: SrcSpanInfo -> Int
```

Comment:
From here on down looks unnecessary.  It's obsolete because of other suggested changes above.